hail/src/main/scala/is/hail/expr/ir/ForwardRelationalLets.scala

```diff
@@ -1,81 +1,83 @@
+            } else
+              RelationalLet(name, rewrite(value).asInstanceOf[IR], rewrite(body).asInstanceOf[IR])
+          case RelationalLetTable(name, value, body) =>
+            if (shouldForward(name)) {
+              env(name) = rewrite(value).asInstanceOf[IR]
+              rewrite(body)
+            } else RelationalLetTable(
+              name,
+              rewrite(value).asInstanceOf[IR],
+              rewrite(body).asInstanceOf[TableIR],
+            )
+          case RelationalLetMatrixTable(name, value, body) =>
+            if (shouldForward(name)) {
+              env(name) = rewrite(value).asInstanceOf[IR]
+              rewrite(body)
+            } else RelationalLetMatrixTable(
+              name,
+              rewrite(value).asInstanceOf[IR],
+              rewrite(body).asInstanceOf[MatrixIR],
+            )
+          case RelationalLetBlockMatrix(name, value, body) =>
+            if (shouldForward(name)) {
+              env(name) = rewrite(value).asInstanceOf[IR]
+              rewrite(body)
+            } else RelationalLetBlockMatrix(
+              name,
+              rewrite(value).asInstanceOf[IR],
+              rewrite(body).asInstanceOf[BlockMatrixIR],
+            )
+          case x @ RelationalRef(name, _) =>
+            env.getOrElse(name, x)
+          case _ => ir1.mapChildren(rewrite)
+        }
 
-      def recur(ir1: BaseIR): BaseIR = ir1 match {
-        case RelationalLet(name, value, body) =>
-          if (shouldForward(usages(name))) {
-            m(name) = recur(value).asInstanceOf[IR]
-            recur(body)
-          } else RelationalLet(name, recur(value).asInstanceOf[IR], recur(body).asInstanceOf[IR])
-        case RelationalLetTable(name, value, body) =>
-          if (shouldForward(usages(name))) {
-            m(name) = recur(value).asInstanceOf[IR]
-            recur(body)
-          } else RelationalLetTable(
-            name,
-            recur(value).asInstanceOf[IR],
-            recur(body).asInstanceOf[TableIR],
-          )
-        case RelationalLetMatrixTable(name, value, body) =>
-          if (shouldForward(usages(name))) {
-            m(name) = recur(value).asInstanceOf[IR]
-            recur(body)
-          } else RelationalLetMatrixTable(
-            name,
-            recur(value).asInstanceOf[IR],
-            recur(body).asInstanceOf[MatrixIR],
-          )
-        case RelationalLetBlockMatrix(name, value, body) =>
-          if (shouldForward(usages(name))) {
-            m(name) = recur(value).asInstanceOf[IR]
-            recur(body)
-          } else RelationalLetBlockMatrix(
-            name,
-            recur(value).asInstanceOf[IR],
-            recur(body).asInstanceOf[BlockMatrixIR],
-          )
-        case x @ RelationalRef(name, _) =>
-          m.getOrElse(name, x)
-        case _ => ir1.mapChildren(recur)
-      }
+        val ir = rewrite(ir0)
 
-      recur(ir0)
+        try
+          TypeCheck(ctx, ir)
+        catch {
```

Comment:
Moving typecheck here also to better mirror `ForwardLets`.
Probably these two optimisations can be combined.